A New Data Structure - aka AoS Reminders v4.0.0 (#1074)
daviseford/aos-reminders: DATA_STRUCTURE.md

- [Top Tier: Factions, Sub-Factions, and Flavors](#top-tier-factions-sub-factions-and-flavors)
  - [Glossary](#glossary)
    - [Faction](#faction)
    - [Sub-Factions](#sub-factions)
      - [Base Sub-Faction](#base-sub-faction)
    - [Flavors](#flavors)
  - [Examples](#examples)

# Top Tier: Factions, Sub-Factions, and Flavors

## Glossary

### Faction

First, let's define a **Faction**

- A **Faction** is the highest-level representation of a army/faction in AoS.
- A **Faction** is the aggregate set of rules that govern many **Sub-Factions**. 
- If a **Faction** only has one **Sub-Faction**, that will be considered the **Base Sub-Faction**

>  - A **Faction** must have a **Base Sub-Faction** 

Reviewer: I think this base sub-faction stuff is confusing.
I initially found it so, and think I should have gone with my gut.

When I first read "base" I was expecting inheritance, and this part of its functionality is borne out by the ability to make it "not visible".
But we're also mixing this with the need for a default sub-faction, and therefore making a "base" sub-faction even when there's no inheritance and much more of a sibling relationship or, worse, we'd have a situation where we have a non-visible parent sub-faction for inheritance purposes (eg "BaseLegions") and then one of its children would be defined as the "base sub-faction" (eg "Grand Host of Nagash").

So I reckon we should split these two behaviours out, leading to something like:

A **Faction** must have at least one **Sub-Faction**
A **Faction** must have a **Default Sub-Faction**
Any Sub-Faction can be extended by other Sub-Factions
Any Sub-Faction can make itself as **Not visible** if it exists only to be extended by other Sub-Factions; by convention this would have something like "Base" in its name to indicate its purpose

Author: > A **Faction** must have at least one **Sub-Faction**
> A **Faction** must have a **Default Sub-Faction**
> Any **Sub-Faction** can be extended by other **Sub-Factions**
> Any **Sub-Faction** can mark itself as **Not visible** if it exists only to be extended by other Sub-Factions; by convention this would have something like "Base" in its name to indicate its purpose

agreed 100%

- If a **Faction** has multiple **Sub Factions**, 
  - It must assign one as a **Base Sub-Faction**
 - And the user will be allowed to select a variant

### Sub-Factions

Next, **Sub-Factions**.

> - A Sub-Faction belongs to a **Faction**

Author: I think Lethisian Defenders would be a Faction

### Sub-Factions

Next, **Sub-Factions**.

- A Sub-Faction belongs to a **Faction**
- A **Sub-Faction** can dictate:
  - What **Flavors** it may contain
  - What units it has
  - What battalions it has
- A **Sub-Faction** may also apply certain Abilities or effects, units, spells, etc

#### Default Sub-Faction

A **Sub-Faction** may be a **Default Sub-Faction**

~~A **Base Sub-Faction** can mark itself as **Not Visible**~~

We generally will try to extend other **Sub-Factions** from the **Base Sub-Faction**

  ~~- For example - `Ironjawz` and `Bonesplitterz` (**Sub-Factions**) share a lot of data from the `Orruk Warclans` (**Faction**) book.~~
  ~~- So we would want three **Sub-Factions** in total:~~
    ~~- **Base Sub-Faction** - marked as **Not Visible**~~
    ~~- **`Ironjawz` Sub-Faction**~~
    ~~- **`Bonesplitterz` Sub-Faction**~~

For example, `Stormcast` would probably be the **Default Sub-Faction** for the **STORMCAST** **Faction**, and then `Stormcast Stormkeep` could potentially extend it and modify it.

### Flavors

Lastly in this high-level tier (but at the bottom), **Flavors**

- These are the more "traditional" ways of configuring **Factions**. 

> - An **Allegiance** may belong to a one or many **Sub-Factions**

Reviewer: ```suggestion
- A **Flavor** may belong to a one or many **Sub-Factions**
```

> - An **Allegiance** can dictate:

Reviewer: ```suggestion
- A **Flavor** can dictate:
```

- What units/battalions it has

> - An **Allegiance** can also apply Abilities or effects units, spells, etc

Reviewer: ```suggestion
- A **Flavor** can also apply Abilities or effects units, spells, etc
```
